# Buổi 01 | 10/04/2021

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1hUnQUC0WkMPRI22RV9urMzvLQBxbEgBD?usp=sharing)

#Data structures

In [ ]:
a = True
type(a)

bool

In [ ]:
a = 10 
b = 3
print(divmod(a,b))
print(a**b)

(3, 1)
1000


In [ ]:
import math

In [ ]:
lst = ['apples', 1, 'oranges']
lst[0] = 2
lst

[2, 1, 'oranges']

In [ ]:
# cac phan tu trong tuple khong sua duoc (immutable)
tup = ('apples', 1, 'oranges')
tup[0] = 2
tup

TypeError: ignored

In [ ]:
x = range(0, 50, 5)
list(x)

[0, 5, 10, 15, 20, 25, 30, 35, 40, 45]

In [ ]:
s = '"Pham Duc The"'
print(s)

"Pham Duc The"


In [ ]:
s = "I'm The"
print(s)

I'm The


In [ ]:
s = """Pham
Duc 
The"""
print(s)

Pham
Duc 
The


In [ ]:
lst = [1, [1,2,3], ['a', 'b', 'c'], (4,5,6)]
print(lst)
print(lst[2])
print(lst[2][2])

[1, [1, 2, 3], ['a', 'b', 'c'], (4, 5, 6)]
['a', 'b', 'c']
c


In [ ]:
Name = "The, Cuong, Hung, Kien, Le, Quy, Nhan"
print(Name[0:-1])
print(Name[:])
print(Name[:15])
print(Name[15:])
print(Name[0:7])
print(Name[5:15])
print(Name[-15:-5])
print(Name[-1])

The, Cuong, Hung, Kien, Le, Quy, Nha
The, Cuong, Hung, Kien, Le, Quy, Nhan
The, Cuong, Hun
g, Kien, Le, Quy, Nhan
The, Cu
Cuong, Hun
, Le, Quy,
n


In [ ]:
Name = ["The", "Cuong", "Hung", "Kien", "Le", "Quy", "Nhan"]
print(Name[0:-1])
print(Name[:])
print(Name[:5])
print(Name[5:])
print(Name[0:7])
print(Name[1:5])
print(Name[-5:-2])
print(Name[-1])

['The', 'Cuong', 'Hung', 'Kien', 'Le', 'Quy']
['The', 'Cuong', 'Hung', 'Kien', 'Le', 'Quy', 'Nhan']
['The', 'Cuong', 'Hung', 'Kien', 'Le']
['Quy', 'Nhan']
['The', 'Cuong', 'Hung', 'Kien', 'Le', 'Quy', 'Nhan']
['Cuong', 'Hung', 'Kien', 'Le']
['Hung', 'Kien', 'Le']
Nhan


In [ ]:
lst = [1,2,3,4,5,6,7,8,9,10]
print(lst[::2])

[1, 3, 5, 7, 9]


In [ ]:
a = [1,2,3]
b = [1,2,3]
print(a==b)
print(a is b)

True
False


In [ ]:
b = a
print(a is b)

True


In [ ]:
b[1] = 10
print(a)

[1, 10, 3]


In [ ]:
b = a[:]
print(a)
print(b)

[1, 10, 3]
[1, 10, 3]


In [ ]:
a[0] = 5
print(a)
print(b)

[5, 10, 3]
[1, 10, 3]


In [ ]:
s = 'Pham Duc The'
print(s*3)
print(len(s))
print(s.index('u'))
print(s.count('h'))
print(max(s))
print(min(s))

Pham Duc ThePham Duc ThePham Duc The
12
6
2
u
 


In [ ]:
1 in (1,2,3)

True

In [ ]:
lst = [[]]*3
print(lst)

[[], [], []]


In [ ]:
lst[0].append(5)
print(lst)

[[5], [5], [5]]


In [ ]:
lst = lst *2
print(lst)

[[5], [5], [5], [5], [5], [5]]


In [ ]:
Name = ["The", "Cuong", "Hung", "Kien", "Le", "Quy", "Nhan"]
Name.sort()
print(Name)

#How to sort by score?
student = [("Nam", 10), ("Minh", 4), ("Xuan", 8), ("My", 2)]
student.sort(key=lambda x: x[1])
print(student)

student.sort(key=lambda x: x[1], reverse=True)
print(student)

['Cuong', 'Hung', 'Kien', 'Le', 'Nhan', 'Quy', 'The']
[('My', 2), ('Minh', 4), ('Xuan', 8), ('Nam', 10)]
[('Nam', 10), ('Xuan', 8), ('Minh', 4), ('My', 2)]


In [ ]:
name = 'Pham Duc The'

In [ ]:
name.count('h')

2

In [ ]:
name.find('a')

2

In [ ]:
name.find('K')

-1

In [ ]:
name.replace('e', 'e Pro')

'Pham Duc The Pro'

In [ ]:
name.upper()

'PHAM DUC THE'

In [ ]:
name.split(' ')

['Pham', 'Duc', 'The']

In [ ]:
'-'.join(name)

'P-h-a-m- -D-u-c- -T-h-e'

In [ ]:
a = [1,2,3,4,5,6]
''.join(a)

TypeError: ignored

In [ ]:
''.join(map(str,a))

'123456'

In [ ]:
c = dict(zip(['one', 'two', 'three'], [1, 2, 3]))
c

{'one': 1, 'three': 3, 'two': 2}

In [ ]:
a = 'Pham'
b = 'Duc'
c = 'The'
print(a + ' ' + b, c)

Pham Duc The


In [ ]:
print('{} {} {}'.format(a,b,c))

Pham Duc The


In [ ]:
print('{0} {1} cost ${2}'.format(6, 'bananas', 1.74))

6 bananas cost $1.74


In [ ]:
print('{a} {b} cost ${c}'.format(b='bananas', a=6, c=1.74))

6 bananas cost $1.74


In [ ]:
'{0}/{1}/{2}'.format('Pham', 'Duc', 'The')

'Pham/Duc/The'

In [ ]:
'{2}.{1}.{0}/{0}{0}.{1}{1}.{2}{2}'.format('Pham', 'Duc', 'The')

'The.Duc.Pham/PhamPham.DucDuc.TheThe'

In [ ]:
'{}/{}/{}'.format('Pham', 'Duc', 'The')

'Pham/Duc/The'

In [ ]:
'{}{}'.format('Pham', 'Duc', 'The')

'PhamDuc'

# Conditions, Branching & Loops

In [ ]:
for x in range(1,5):
  print(str(x).rjust(2),
        str(x**2).rjust(3),
        str(x**3).rjust(3))

 1   1   1
 2   4   8
 3   9  27
 4  16  64


In [ ]:
for x in range(1,5):
  print("{:2d} {:3d} {:3d}".format(x, x**2, x**3))

 1   1   1
 2   4   8
 3   9  27
 4  16  64


In [ ]:
print("{:03d} {:05.2f}".format(12, 12.5435345))

012 12.54


In [ ]:
print("{:*<10}".format("Hello"))

Hello*****


In [ ]:
print("{:*>10}".format("Hello"))

*****Hello


In [ ]:
print("{:->10}".format("Hello"))

-----Hello


In [ ]:
print("{:-^100}".format("Hello"))

-----------------------------------------------Hello------------------------------------------------


In [ ]:
a=10
b=15
print(f'{a} + {b} = {a + b}')

10 + 15 = 25


In [ ]:
s = 'k'

In [ ]:
s.islower()

True

In [ ]:
n = int(input())
m = int(input())

# StarCode
s = {value for value in range(n+1)}
s.remove(m)
result = ""
for i in s:
  result = result + str(i)*2
      
print(result)
# print(result)

5
2
0011334455


In [ ]:
s1 = input()
s2 = input()

lst1 = []
lst2 = []
result = ''

lst1 = s1.split(',') 
lst2 = s2.split(',') 

for i in range(len(lst1)):
  if int(lst1[i]) > int(lst2[i]):
    result = result + lst1[i] + ',' 
  else:
    result = result + lst2[i] + ',' 

result = result[:-1]
print(result)

1,2,3
2,2,4
['1', '2', '3']
['2', '2', '4']
['2', '2', '4']


In [ ]:
dic = {'Object': 'Python', 'Mark': 10, 'Student': 'NguyenVanNam', 'Male': True, 'City': 'HoChiMinh'}
print(dic)
dic['Male'] = False
print(dic)
print(dic.keys())
print(dic.values())

{'Object': 'Python', 'Mark': 10, 'Student': 'NguyenVanNam', 'Male': True, 'City': 'HoChiMinh'}
{'Object': 'Python', 'Mark': 10, 'Student': 'NguyenVanNam', 'Male': False, 'City': 'HoChiMinh'}
dict_keys(['Object', 'Mark', 'Student', 'Male', 'City'])
dict_values(['Python', 10, 'NguyenVanNam', False, 'HoChiMinh'])


#Bài Tập

##ViPy_021_Xử lý tiếng Việt

Bạn Nam đang gõ văn bản thì bàn phím bị lỗi không thể gõ được dấu cách. Nam khắc phục bằng cách viết thêm dấu | giữ mỗi từ. Việc của bạn là hãy giúp Nam:

Bỏ dấu | và thêm dấu cách vào giữa các từ
Chuyễn các chữ hoa thành chữ thường.
Ví dụ:

Input

Nam|Hom|Nay|Bi|Om.|Nam|Muon|Xin|Phep|Co|Nghi|Hoc|7|Ngay.

Output:

nam hom nay bi om. nam muon xin phep co nghi hoc 7 ngay.

Input Format

.

Constraints

.

Output Format

.

In [ ]:
#!/bin/python3

import math
import os
import random
import re
import sys



if __name__ == '__main__':
    input_data = input()
    result = input_data.replace('|', ' ')
    result = result.lower()
    print(result)

Nam|Hom|Nay|Bi|Om.|Nam|Muon|Xin|Phep|Co|Nghi|Hoc|7|Ngay.
nam hom nay bi om. nam muon xin phep co nghi hoc 7 ngay.


##ViPy_022_Dict

Nhập 1 dict có giá trị như sau:

Object: 'Python'

Mark: 10

Student: 'NguyenVanNam'

Male: True

City: 'HoChiMinh'

+ Xuất dữ liệu đó ra màn hình
+ Chuyển giới tính Male = False xuất ra màn hình
+ Xuất màn hình danh sách các keys
+ Xuất màn hình danh sách cac value

Input Format

.

Constraints

.

Output Format

.

In [ ]:
#!/bin/python3

import math
import os
import random
import re
import sys



if __name__ == '__main__':
    # Write your code here
    dic = {'Object': 'Python', 'Mark': 10, 'Student': 'NguyenVanNam', 'Male': True, 'City': 'HoChiMinh'}
    print(dic)
    dic['Male'] = False
    print(dic)
    print(dic.keys())
    print(dic.values())

{'Object': 'Python', 'Mark': 10, 'Student': 'NguyenVanNam', 'Male': True, 'City': 'HoChiMinh'}
{'Object': 'Python', 'Mark': 10, 'Student': 'NguyenVanNam', 'Male': False, 'City': 'HoChiMinh'}
dict_keys(['Object', 'Mark', 'Student', 'Male', 'City'])
dict_values(['Python', 10, 'NguyenVanNam', False, 'HoChiMinh'])


##ViPy_022_Set

Nhập 2 số n, m. Thực hiện các thao tác sau:

+ Tạo 1 Set chứa các giá trị chạy từ 0 -> n (0, 1, 2, ..., n)
+ Xoá phần từ có giá trị m trong Set
+ Từ set này hãy tạo 1 chuỗi string bao gồm các chữ số được lặp lại 2 lần theo thứ tự tăng dần.
+ Xuất ra màn hình kết quả nhận được.
Ví dụ:

n = 5, m = 2

Set = {0, 1, 2, 3, 4, 5}

Xoá phần tử 2

Set = {0, 1, 3, 4, 5}
Xuất ra màn hình chuỗi

'0011334455'

Input Format

.

Constraints

.

Output Format

.

In [ ]:
#!/bin/python3

import math
import os
import random
import re
import sys



if __name__ == '__main__':
    n = int(input().strip())
    m = int(input().strip())

    # StarCode
    s = {value for value in range(n+1)}
    s.remove(m)
    result = ""
    for i in s:
      result = result + str(i)*2

    print(result)
    # print(result)


5
2
0011334455


##ViPy_031_Tìm số lớn nhất

Nhập 2 số float a, b

Xuất ra số lớn nhất. Nếu a = b xuất ra - 1

Input Format

.

Constraints

.

Output Format

.

In [ ]:
#!/bin/python3

import math
import os
import random
import re
import sys



if __name__ == '__main__':
    # Write your code here
    a = float(input())
    b = float(input())
    
    if a > b:
        print(a)
    elif a < b:
        print(b)
    else:
     print(-1)
        


3
4
4.0


##ViPy_032_Tổng số lẻ

Cho 2 số nguyên a và b được nhập từ bàn phím, bạn hãy viết chương trình hiển thị ra tổng các số lẻ từ a tới b.

Ví dụ nếu bạn nhập a = 3, b = 9 thì màn hình sẽ hiển thị ra:

24

Giải thích: 3 + 5 + 7 + 9 = 24`.

Đầu vào luôn đảm bảo b > a.

Input Format

.

Constraints

.

Output Format

.

In [ ]:
#!/bin/python3

import math
import os
import random
import re
import sys



if __name__ == '__main__':
    a = int(input().strip())

    b = int(input().strip())

    # Write your code here
    s = 0
    for i in range(a,b+1):
        if i % 2 != 0 :
            s += i
    
    print(s)


3
9
24


##ViPy_033_List comprehension

Nhập vào 1 số n

Sử dụng list comprehension để tạo 1 list bao gồm các số nguyên chia hết cho n lớn hơn 0 và bé hơn 100

In ra màn hình list đó

Input Format

.

Constraints

.

Output Format

.

In [ ]:
#!/bin/python3

import math
import os
import random
import re
import sys



if __name__ == '__main__':
    # Write your code here
    n = int(input())
    lst = []
    for i in range(1,100):
        if i % n == 0:
            lst.append(i)
            
    print(lst)


15
[15, 30, 45, 60, 75, 90]


##ViPy_033_sum

Nhập số n

In ra màn hình các cặp giá trị a b. Trong đó

0<= a, b <=n

a + b = int(1.5 * n)

Ví dụ:

Input:

20

Output:

10 20

11 19

12 18

13 17

14 16

15 15

16 14

17 13

18 12

19 11

20 10

Input Format

.

Constraints

.

Output Format

.

In [ ]:
#!/bin/python3

import math
import os
import random
import re
import sys



if __name__ == '__main__':
    # Write your code here
    n = int(input())
    m = int(1.5*n)
    b = n
    a = m - b
    
    while abs(a)+abs(b)==m:
        print(a, b)
        a += 1
        b -= 1
        if a > n or b > n:
            break

20
10 20
11 19
12 18
13 17
14 16
15 15
16 14
17 13
18 12
19 11
20 10


##ViPy_041_Print string

Cho chuỗi s được nhập vào từ bàn phím, bạn hãy viết chương trình Python để đếm số các chữ cái in hoa và in thường trong chuỗi và in kết quả ra màn hình theo từng dòng như sau:

Given string: {P1}

Number of uppercase letters: {P2}

Number of lowercase letters: {P3}

Trong đó:

+ {P1} là chuỗi s được nhập vào từ bàn phím.
+ {P2} là số các chữ cái in hoa.
+ {P3} là số các chữ cái in thường.

Ví dụ:

Nếu bạn nhập You Are Apple In My Eye thì màn hình sẽ hiển thị ra:

Given string: You Are Apple In My Eye

Number of uppercase letters: 6

Number of lowercase letters: 12

Input Format

.

Constraints

.

Output Format

.

In [ ]:
s = input()

Count_Upper = 0
Count_Lower = 0

for i in range(0, len(s)):
    if s[i].islower():
        Count_Lower += 1
    elif s[i].isupper():
        Count_Upper += 1
    else:
        continue
        
print('Given string: {}'.format(s))
print('Number of uppercase letters: {}'.format(Count_Upper))
print('Number of lowercase letters: {}'.format(Count_Lower))

You Are Apple In My Eye
Given string: You Are Apple In My Eye
Number of uppercase letters: 6
Number of lowercase letters: 12


##ViPy_042_max

Nhập 2 string bao gồm danh sách các số nguyên cách nhau bởi dấu ,

Số lượng số nguyên ở 2 danh sách là như sau.

In ra màn hình danh sách mới là số lớn nhất trong 2 danh sách theo từng thứ tự phần tử.

Ví dụ

Input

1,3,5,5,7,4

4,6,2,1,6,3

Outputt

4,6,5,5,7,4

Input Format

.

Constraints

.

Output Format

.

In [ ]:
s1 = input()
s2 = input()

lst1 = []
lst2 = []
result = ''

lst1 = s1.split(',') 
lst2 = s2.split(',') 

for i in range(len(lst1)):
  if int(lst1[i]) > int(lst2[i]):
    result = result + lst1[i] + ',' 
  else:
    result = result + lst2[i] + ',' 

result = result[:-1]
print(result)

1,3,5,5,7,4
4,6,2,1,6,3
4,6,5,5,7,4
